### DISCUSSION TITLE
Automapper on HotChocolate Optional types

### DISCUSSION THREAD

**Discussion by shidcordero** [LuckyPennySoftware/AutoMapper#3925] [NA]
### Source/destination types

```csharp
// Source
public class TagInput
{
    public int? Id { get; set; }
    public Optional<string> Name { get; set; }
    public Optional<int?> TagTypeId { get; set; }
}

// Destination
public class Tag 
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public int Id { get; set; }
    public string Name { get; set; }
    public int TagTypeId { get; set; }
    public virtual TagType TagType { get; set; }
}
```

### Mapping configuration

```csharp
public class TagInputProfile : Profile
{
    public TagInputProfile()
    {
        CreateMap<TagInput, Tag>()
            .ForMember(dest => dest.Id, opt => opt.Condition(src => src.Id.HasValue))
            .ForMember(dest => dest.Name, opt => {
                opt.PreCondition(src => src.Name.HasValue);
                opt.MapFrom(src => src.Name.Value);
            })
            .ForMember(dest => dest.TagTypeId, opt => {
                opt.PreCondition(src => src.TagTypeId.HasValue);
                opt.MapFrom(src => src.TagTypeId.Value);
            })
            .EqualityComparison((d, s) => d.Id == s.Id)
            .ReverseMap();
    }
}
```

### Version: 
Automapper: 8.1.1
Automapper.Collection: 7.0.1
HotChocolate: 12.6.1
EF6

### Expected behavior

To be able to map Optional<T> type to the destination. Right now it just keeps throwing this error on run:
```
System.ArgumentException: 'Expression of type 'System.Int32' cannot be used for parameter of type 'System.Nullable`1[System.Int32]' of method 'HotChocolate.Optional`1[System.Nullable`1[System.Int32]] op_Implicit(System.Nullable`1[System.Int32])' (Parameter 'arg0')'
```

Note that only Nullable<T> does not work. The string type is working since I did not enable nullable reference types. Even nullable reference type was enabled, its still throwing the same error. I am not sure what's causing the error. I actually expect the above mapping code to work since the precondition will only map those if `.HasValue` (if `.HasValue` is true of Optional, it means that the user pass a value on that field. Optional is used for partial updates). Once the precondition is valid, it will only map it to the value(for example if its `Optional<int?>` and `.HasValue` is true, it will map from `.Value` of `Optional<int?>` that means its `int?`).

### Actual behavior

It throws the error above.

### Steps to reproduce

Not sure if this is the correct step to reproduce but the error was thrown when `new MapperConfiguration()` was called on Startup.cs
```csharp
var configuration = new MapperConfiguration(config =>
{
    config.AddCollectionMappers();
    config.AddMaps("Horsebee.API");
});

services.AddSingleton(sp => configuration.CreateMapper());
```

**Comment by lbargaoanu** [comment]
A repro would help. Make a [gist](https://gist.github.com/lbargaoanu/9c7233441c3a3413cc2b9b9ebb5964a9) that we can execute and see fail. That means the only reference must be to the latest AM. _Everything_ else must be part of the gist.

**Reply by shidcordero** [reply]
here's a dotnet fiddle to replicate the issue https://dotnetfiddle.net/pmXTRE
 Here's the [gist](https://gist.github.com/shidcordero/93c6cf83562074c84d856ee6742da766) also

**Reply by lbargaoanu** [reply]
That works with the latest AM.

**Reply by shidcordero** [reply]
@lbargaoanu it works great. Now the only issue I have is with Automapper.Collections still not supporting latest version of Automapper.
